lib/puppetserver/ca/x509_loader.rb

```diff
@@ -0,0 +1,123 @@
+require 'openssl'
+
+module Puppetserver
+  module Ca
+    # Load, validate, and store x509 objects needed by the Puppet Server CA.
+    class X509Loader
+
+      attr_reader :errors, :certs, :key, :crls
+
+      def initialize(bundle_path, key_path, chain_path = nil)
+        @errors = []
+
+        @certs = load_certs(bundle_path)
+        @key = load_key(key_path)
+        @crls = chain_path ? load_crls(chain_path) : []
+
+        validate(@certs, @key, @crls)
+      end
+
+      # Only do as much validation as is possible, assume whoever tried to
+      # load the objects wrote errors about any invalid ones, but that bundle
+      # and chain may be empty arrays and pkey may be nil.
+      def validate(bundle, pkey, chain)
+        if !chain.empty? && !bundle.empty?
```

Comment:
I never know if I should put this logic (like if nil ...don't call this) here or in the method itself.   Do you think its better to make the methods "nil safe" or to not call them if the arguments are nil?

Reply:
I went back and forth on that here myself. I think that having the caller do the nil-checks would essentially inline this method into the caller (the initialize).

The validation seemed sufficiently different to me to warrant being in a different method, but... I'm not sure they're really a different level of abstraction than the loading that currently happens in the initialize method....

Reply:
I like putting all of this under the `validate` banner. If Amy was referring to whether to move this line down inside the subsequent method, in this case I would keep it outside, since the `validate_crl_and_cert` method cares about something more specific than the whole bundle and chain. In order to give it only what it needs, we need to make sure these things aren't empty first.